python/lsst/daf/butler/datastores/chainedDatastore.py

```diff
@@ -0,0 +1,342 @@
+        """Check if the dataset exists in one of the datastores.
+
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            Reference to the required dataset.
+
+        Returns
+        -------
+        exists : `bool`
+            `True` if the entity exists in one of the child datastores.
+        """
+        for datastore in self.datastores:
+            if datastore.exists(ref):
+                log.debug("Found %s in datastore %s", ref, datastore.name)
+                return True
+        return False
+
+    def get(self, ref, parameters=None):
+        """Load an InMemoryDataset from the store.
+
+        The dataset is returned from the first datastore that has
+        the dataset.
+
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            Reference to the required Dataset.
+        parameters : `dict`
+            `StorageClass`-specific parameters that specify, for example,
+            a slice of the Dataset to be loaded.
+
+        Returns
+        -------
+        inMemoryDataset : `object`
+            Requested Dataset or slice thereof as an InMemoryDataset.
+
+        Raises
+        ------
+        FileNotFoundError
+            Requested dataset can not be retrieved.
+        TypeError
+            Return value from formatter has unexpected type.
+        ValueError
+            Formatter failed to process the dataset.
+        """
+
+        for datastore in self.datastores:
+            try:
+                inMemoryObject = datastore.get(ref, parameters)
+                log.debug("Found Dataset %s in datastore %s", ref, datastore.name)
+                return inMemoryObject
+            except FileNotFoundError:
+                pass
+
+        raise FileNotFoundError("Dataset {} could not be found in any of the datastores".format(ref))
+
+    def put(self, inMemoryDataset, ref):
+        """Write a InMemoryDataset with a given `DatasetRef` to each
+        datastore.
+
+        Parameters
+        ----------
+        inMemoryDataset : `object`
+            The Dataset to store.
+        ref : `DatasetRef`
+            Reference to the associated Dataset.
+
+        Raises
+        ------
+        TypeError
+            Supplied object and storage class are inconsistent.
+        """
+
+        for datastore in self.datastores:
+            datastore.put(inMemoryDataset, ref)
+
+        if self._transaction is not None:
+            self._transaction.registerUndo('put', self.remove, ref)
+
+    def ingest(self, *args, **kwargs):
+        """Add an on-disk file with the given `DatasetRef` to the store,
+        possibly transferring it.
+
+        This method is forwarded to each of the chained datastores, trapping
+        cases where a datastore has not implemented file ingest and ignoring
+        them.
+
+        A transfer mode of None is not supported since that requires the
+        file to have been previously copied to each individual datastore.
```

Comment:
Seems like `None` ought to be okay if only one `Datastore` in the chain supports ingest, and the file is already in that one.  I think you'll get the right behavior if you just remove that special check.

Reply:
My real problem here was to understand how a file turned up in one of the data stores in the first place. The test infrastructure doesn't actually know that a chained datastore is being used so I wasn't sure how much special casing I should be including to copy that file in to the right place during the tests (it would have to know to copy the test file to `.datastores[1].root`). I now realize that I should add a second posix datastore to the chaining tests since that would demonstrate that ingest with transfer=move would fail on the second datastore. Then we have to decide whether a single ingest working is good enough or if all implemented ingests should work (transfer=move could work if the chained ingest copied the file itself to a temporary area so that each datastore could move it itself)

Reply:
What about making it so `ingest` is only supported if exactly one `Datastore` accepts the file?  I can't think of a good use case for anything more complex than that (though we would probably want to let different `Datastores` in the chain accept different files).

Reply:
Hmm. It seems a bit odd to me that ingest would stop at the first success (or continue trying in order to ensure that they all subsequently failed) but `put` sends to each child datastore.